# 2.4. Evaluating Hyperparameters _ CNN Model

## The following script contains the following:
### 1. Import data libraries and data sets
### 2. Preprocessing the data frames
### 3. Bayesian optimization function 
### 4. Build and run CNN keras models
### 5. Run confusion matrix

## 01. Import data libraries and data sets

In [1]:
# Import libraries.

import pandas as pd
import numpy as np
import seaborn as sns
import os
import operator
import time
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
import tensorflow as tf
from numpy import unique
from numpy import reshape
import tensorflow.keras as keras
from keras.models import Sequential
from sklearn.model_selection import cross_val_score
from keras.layers import Conv1D, Conv2D, Dense, BatchNormalization, Flatten, MaxPooling1D, Dropout
from tensorflow.keras.utils import to_categorical
from keras.optimizers import Adam, SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from keras.callbacks import EarlyStopping, ModelCheckpoint
from scikeras.wrappers import KerasClassifier
from math import floor
from sklearn.metrics import make_scorer, accuracy_score
from bayes_opt import BayesianOptimization
from sklearn.model_selection import StratifiedKFold
from keras.layers import LeakyReLU
LeakyReLU = LeakyReLU(negative_slope=0.1)
import warnings

In [2]:
# Supress scientific notation for easier analysis profiling
pd.set_option('display.float_format', '{:.2f}'.format)

# Set display options to show all columns without truncation
pd.set_option('display.max_columns', None)

In [3]:
# Create path
path = r'C:\Users\Quinn\Documents\CF - Data Analysis\Machine Learning\ClimateWins'

In [4]:
# Import the dataset
climate = pd.read_csv(os.path.join(path, '02 Data Sets', 'cleaned_climate.csv'))
climate.head(5)

Unnamed: 0      DATE  MONTH  BASEL_cloud_cover  BASEL_humidity  \
0           0  19600101      1                  7            0.85   
1           1  19600102      1                  6            0.84   
2           2  19600103      1                  8            0.90   
3           3  19600104      1                  3            0.92   
4           4  19600105      1                  6            0.95   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0            1.02                    0.32                 0.09   
1            1.02                    0.36                 1.05   
2            1.02                    0.18                 0.30   
3            1.02                    0.58                 0.00   
4            1.02                    0.65                 0.14   

   BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  BASEL_temp_max  \
0            0.70             6.50            0.80           10.90   
1            1.10             6.10            3.30           10.10   
2            0.00             8.50            5.10            9.90   
3            4.10             6.30            3.80           10.60   
4            5.40             3.00           -0.70            6.00   

   BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0                     1               0.81               1.02   
1                     6               0.84               1.02   
2                     6               0.77               1.02   
3                     8               0.93               1.03   
4                     8               0.99               1.03   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                       0.88                    0.00               7.00   
1                       0.25                    0.00               0.00   
2                       0.67                    0.00               3.50   
3                       0.25                    0.00               0.00   
4                       0.25                    0.06               0.00   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0                3.70              -0.90               7.90   
1                2.90               2.20               4.40   
2                3.10              -0.50               6.40   
3                2.00              -2.00               3.00   
4                2.00               0.70               2.80   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0                     4               0.67               1.02   
1                     4               0.67               1.02   
2                     4               0.67               1.02   
3                     4               0.67               1.02   
4                     4               0.67               1.02   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                       0.44                    0.01               2.30   
1                       0.18                    0.31               0.00   
2                       0.30                    0.00               0.60   
3                       0.19                    0.00               0.00   
4                       0.19                    0.00               0.00   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0                2.40              -0.40               5.10   
1                2.30               1.40               3.10   
2                2.70               1.70               5.30   
3                2.00               0.40               4.40   
4                2.50               1.10               5.30   

   DEBILT_cloud_cover  DEBILT_humidity  DEBILT_pressure  \
0                   7             0.85             1.00   
1                   8             0.90             1.01   
2                   6             0.92             1.02   
3                   8             0.95             1.03   
4                   6            

In [5]:
pleasant = pd.read_csv(os.path.join(path, '02 Data Sets', 'Dataset-Answers-Weather_Prediction_Pleasant_Weather.csv'))
pleasant.head(5)

DATE  BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0  19600101                       0                          0   
1  19600102                       0                          0   
2  19600103                       0                          0   
3  19600104                       0                          0   
4  19600105                       0                          0   

   BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                          0                        0   
1                          0                        0   
2                          0                        0   
3                          0                        0   
4                          0                        0   

   DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                            0                          0   
1                            0                          0   
2                            0                          0   
3                            0                          0   
4                            0                          0   

   KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                        0                           0   
1                        0                           0   
2                        0                           0   
3                        0                           0   
4                        0                           0   

   MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   

   MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                          0                      0   
1                          0                      0   
2                          0                      0   
3                          0                      0   
4                          0                      0   

   SONNBLICK_pleasant_weather  STOCKHOLM_pleasant_weather  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   

   VALENTIA_pleasant_weather  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0

## 02. Preprocessing data frames

In [6]:
# Drop the "DATE , MONTH" columsn
df = climate.drop(columns = ['DATE','MONTH','Unnamed: 0'])
df

BASEL_cloud_cover  BASEL_humidity  BASEL_pressure  \
0                      7            0.85            1.02   
1                      6            0.84            1.02   
2                      8            0.90            1.02   
3                      3            0.92            1.02   
4                      6            0.95            1.02   
...                  ...             ...             ...   
22945                  1            0.79            1.02   
22946                  6            0.77            1.02   
22947                  4            0.76            1.02   
22948                  5            0.80            1.02   
22949                  5            0.84            1.02   

       BASEL_global_radiation  BASEL_precipitation  BASEL_sunshine  \
0                        0.32                 0.09            0.70   
1                        0.36                 1.05            1.10   
2                        0.18                 0.30            0.00   
3                        0.58                 0.00            4.10   
4                        0.65                 0.14            5.40   
...                       ...                  ...             ...   
22945                    1.34                 0.22            7.70   
22946                    1.34                 0.22            5.40   
22947                    1.34                 0.22            6.10   
22948                    1.34                 0.22            5.80   
22949                    1.34                 0.22            3.20   

       BASEL_temp_mean  BASEL_temp_min  BASEL_temp_max  BELGRADE_cloud_cover  \
0                 6.50            0.80           10.90                     1   
1                 6.10            3.30           10.10                     6   
2                 8.50            5.10            9.90                     6   
3                 6.30            3.80           10.60                     8   
4                 3.00           -0.70            6.00                     8   
...                ...             ...             ...                   ...   
22945            15.90           11.40           21.40                     2   
22946            16.70           14.30           21.90                     0   
22947            16.70           13.10           22.40                     2   
22948            15.40           11.60           21.10                     1   
22949            13.50            9.90           19.20                     4   

       BELGRADE_humidity  BELGRADE_pressure  BELGRADE_global_radiation  \
0                   0.81               1.02                       0.88   
1                   0.84               1.02                       0.25   
2                   0.77               1.02                       0.67   
3                   0.93               1.03                       0.25   
4                   0.99               1.03                       0.25   
...                  ...                ...                        ...   
22945               0.68               1.03                       1.57   
22946               0.68               1.03                       1.57   
22947               0.68               1.03                       1.57   
22948               0.68               1.02                       1.57   
22949               0.68               1.02                       1.57   

       BELGRADE_precipitation  BELGRADE_sunshine  BELGRADE_temp_mean  \
0                        0.00               7.00                3.70   
1                        0.00               0.00                2.90   
2                        0.00               3.50                3.10   
3                        0.00               0.00                2.00   
4                        0.06               0.00                2.00   
...                       ...                ...                 ...   
22945                    0.18               5.70               18.20   
22946              

In [7]:
# Drop the "DATE" from pleasent days dataset
pleasant = pleasant.drop(columns = ['DATE'])
pleasant

BASEL_pleasant_weather  BELGRADE_pleasant_weather  \
0                           0                          0   
1                           0                          0   
2                           0                          0   
3                           0                          0   
4                           0                          0   
...                       ...                        ...   
22945                       0                          0   
22946                       0                          0   
22947                       0                          0   
22948                       0                          0   
22949                       0                          0   

       BUDAPEST_pleasant_weather  DEBILT_pleasant_weather  \
0                              0                        0   
1                              0                        0   
2                              0                        0   
3                              0                        0   
4                              0                        0   
...                          ...                      ...   
22945                          0                        0   
22946                          0                        0   
22947                          0                        0   
22948                          0                        0   
22949                          0                        0   

       DUSSELDORF_pleasant_weather  HEATHROW_pleasant_weather  \
0                                0                          0   
1                                0                          0   
2                                0                          0   
3                                0                          0   
4                                0                          0   
...                            ...                        ...   
22945                            0                          0   
22946                            0                          0   
22947                            0                          0   
22948                            0                          0   
22949                            0                          0   

       KASSEL_pleasant_weather  LJUBLJANA_pleasant_weather  \
0                            0                           0   
1                            0                           0   
2                            0                           0   
3                            0                           0   
4                            0                           0   
...                        ...                         ...   
22945                        0                           0   
22946                        0                           0   
22947                        0                           0   
22948                        0                           0   
22949                        0                           0   

       MAASTRICHT_pleasant_weather  MADRID_pleasant_weather  \
0                                0                        0   
1                                0                        0   
2                                0                        0   
3                                0                        0   
4                                0                        0   
...                            ...                      ...   
22945                            0                        0   
22946                            0                        0   
22947                            0                        0   
22948                            0                        0   
22949                            0                        0   

       MUNCHENB_pleasant_weather  OSLO_pleasant_weather  \
0                              0                      0   
1                              0                      0   
2                              0                      0   
3                              0  

In [8]:
# Convert df to array.
X = df.to_numpy()

In [9]:
# Reshape df
X = df.values.reshape(-1, 15, 9)

# Ensure the labels are in the same shape as X
y = pleasant.values.reshape(-1, 15)

def _count_classes(y):
    return len(set([tuple(category) for category in y]))

In [10]:
X.shape

(22950, 15, 9)

In [11]:
y.shape

(22950, 15)

In [12]:
#Use argmax to get rid of one-hot encoding and supply the numerical value.
y = np.argmax(y, axis = 1)
print(y.shape)

(22950,)


In [13]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [14]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(16065, 15, 9) (16065,)
(6885, 15, 9) (6885,)


## 03. Bayesian optimization function 

In [15]:
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = 15 #_count_classes(y_train)
# Make scorer accuracy
score_acc = make_scorer(accuracy_score)

In [16]:
# Create function
def bay_area(neurons, activation, kernel, optimizer, learning_rate, batch_size, epochs,
              layers1, layers2, normalization, dropout, dropout_rate): 
    optimizerL = ['SGD', 'Adam', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','SGD']
    optimizerD= {'Adam':Adam(learning_rate=learning_rate), 'SGD':SGD(learning_rate=learning_rate),
                 'RMSprop':RMSprop(learning_rate=learning_rate), 'Adadelta':Adadelta(learning_rate=learning_rate),
                 'Adagrad':Adagrad(learning_rate=learning_rate), 'Adamax':Adamax(learning_rate=learning_rate),
                 'Nadam':Nadam(learning_rate=learning_rate), 'Ftrl':Ftrl(learning_rate=learning_rate)}
    activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
                   'elu', 'exponential', LeakyReLU,'relu']
    
    neurons = round(neurons)
    kernel = round(kernel)
    activation = activationL[round(activation)]
    optimizer = optimizerD[optimizerL[round(optimizer)]]
    batch_size = round(batch_size)
    
    epochs = round(epochs)
    layers1 = round(layers1)
    layers2 = round(layers2)
    
    def cnn_model():
        model = Sequential()
        model.add(Conv1D(neurons, kernel_size=kernel,activation=activation, input_shape=(timesteps, input_dim)))
        #model.add(Conv1D(32, kernel_size=1,activation='relu', input_shape=(timesteps, input_dim)))
        
        if normalization > 0.5:
            model.add(BatchNormalization())
        for i in range(layers1):
            model.add(Dense(neurons, activation=activation)) #(neurons, activation=activation))
        if dropout > 0.5:
            model.add(Dropout(dropout_rate, seed=123))
        for i in range(layers2):
            model.add(Dense(neurons, activation=activation))
        model.add(MaxPooling1D())
        model.add(Flatten())
        model.add(Dense(n_classes, activation='softmax')) #sigmoid softmax
        #model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy']) #categorical_crossentropy
        model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['accuracy']) #categorical_crossentropy
        return model
    es = EarlyStopping(monitor='accuracy', mode='max', verbose=2, patience=20)
    nn = KerasClassifier(build_fn=cnn_model, epochs=epochs, batch_size=batch_size, verbose=2)
    kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=123)
    score = cross_val_score(nn, X_train, y_train, scoring=score_acc, cv=kfold, fit_params={'callbacks':[es]}).mean()
    return score

In [17]:
start = time.time()
params ={
    'neurons': (10, 50),            # Reduced upper limit
    'kernel': (1, 3),
    'activation':(0, 9), 
    'optimizer':(0,7), 
    'learning_rate':(0.001, 0.1),   # Reduced range
    'batch_size': (200, 500),       # Reduced upper limit
    'epochs':(20, 200),              # Reduced upper limit
    'layers1':(1,2),                # Simplified to 1 or 2 layers
    'layers2':(1,2),                # Simplified to 1 or 2 layers
    'normalization':(0,1),
    'dropout':(0,1),
    'dropout_rate':(0,0.3)
}
# Run Bayesian Optimization
nn_opt = BayesianOptimization(bay_area, params, random_state=42)
nn_opt.maximize(init_points=15, n_iter=100)
print('Search took %s minutes' % ((time.time() - start)/60))

|   iter    |  target   | activa... | batch_... |  dropout  | dropou... |  epochs   |  kernel   |  layers1  |  layers2  | learni... |  neurons  | normal... | optimizer |
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------


C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:73: FutureWarning: `fit_params` is deprecated and will be removed in version 1.6. Pass parameters via `params` instead.
  warnings.warn(
C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/48
27/27 - 4s - 149ms/step - accuracy: 0.6263 - loss: 1.7516
Epoch 2/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.2050
Epoch 3/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1830
Epoch 4/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1769
Epoch 5/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1730
Epoch 6/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1715
Epoch 7/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1703
Epoch 8/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1690
Epoch 9/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1690
Epoch 10/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1687
Epoch 11/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1686
Epoch 12/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1680
Epoch 13/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1681
Epoch 14/48
27/27 - 0s - 10ms/step - accuracy: 0.6423 - loss: 1.1678
Epoch 15/48
27/27 - 0s - 10ms/step - accur

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/48
Epoch 1/48


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/48


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/48


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 1         | nan       | 3.371     | 485.2     | 0.732     | 0.1796    | 48.08     | 1.312     | 1.058     | 1.866     | 0.06051   | 38.32     | 0.02058   | 6.789     |
Epoch 1/75
49/49 - 3s - 56ms/step - accuracy: 0.6422 - loss: nan
Epoch 2/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 3/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 4/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 5/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 6/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 7/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 8/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 9/75
49/49 - 0s - 5ms/step - accuracy: 0.6423 - loss: nan
Epoch 10/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 11/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 12/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - loss: nan
Epoch 13/75
49/49 - 0s - 4ms/step - accuracy: 0.6423 - los

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/75


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/75


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/75


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 2         | nan       | 7.492     | 263.7     | 0.1818    | 0.05502   | 74.76     | 2.05      | 1.432     | 1.291     | 0.06157   | 15.58     | 0.2921    | 2.565     |
Epoch 1/127
30/30 - 6s - 195ms/step - accuracy: 0.6422 - loss: 1.0442
Epoch 2/127
30/30 - 0s - 11ms/step - accuracy: 0.7463 - loss: 0.7508
Epoch 3/127
30/30 - 0s - 11ms/step - accuracy: 0.7692 - loss: 0.6781
Epoch 4/127
30/30 - 0s - 11ms/step - accuracy: 0.7846 - loss: 0.6197
Epoch 5/127
30/30 - 0s - 11ms/step - accuracy: 0.8000 - loss: 0.5696
Epoch 6/127
30/30 - 0s - 11ms/step - accuracy: 0.8136 - loss: 0.5225
Epoch 7/127
30/30 - 0s - 11ms/step - accuracy: 0.8264 - loss: 0.4857
Epoch 8/127
30/30 - 0s - 11ms/step - accuracy: 0.8410 - loss: 0.4465
Epoch 9/127
30/30 - 0s - 10ms/step - accuracy: 0.8511 - loss: 0.4085
Epoch 10/127
30/30 - 0s - 11ms/step - accuracy: 0.8570 - loss: 0.4040
Epoch 11/127
30/30 - 0s - 10ms/step - accuracy: 0.8606 - loss: 0.3872
Epoch 12/127
30/30 - 0s - 10ms/step - accuracy: 0.8761 - loss: 0.353

Epoch 115/127
30/30 - 0s - 10ms/step - accuracy: 0.9764 - loss: 0.0663
Epoch 116/127
30/30 - 0s - 10ms/step - accuracy: 0.9791 - loss: 0.0606
Epoch 117/127
30/30 - 0s - 11ms/step - accuracy: 0.9891 - loss: 0.0318
Epoch 118/127
30/30 - 0s - 10ms/step - accuracy: 0.9926 - loss: 0.0232
Epoch 119/127
30/30 - 0s - 11ms/step - accuracy: 0.9777 - loss: 0.0675
Epoch 120/127
30/30 - 0s - 10ms/step - accuracy: 0.9775 - loss: 0.0645
Epoch 121/127
30/30 - 0s - 10ms/step - accuracy: 0.9765 - loss: 0.0690
Epoch 122/127
30/30 - 0s - 10ms/step - accuracy: 0.9817 - loss: 0.0510
Epoch 123/127
30/30 - 0s - 11ms/step - accuracy: 0.9853 - loss: 0.0445
Epoch 124/127
30/30 - 0s - 11ms/step - accuracy: 0.9878 - loss: 0.0366
Epoch 125/127
30/30 - 0s - 10ms/step - accuracy: 0.9968 - loss: 0.0150
Epoch 126/127
30/30 - 0s - 11ms/step - accuracy: 0.9979 - loss: 0.0091
Epoch 127/127
30/30 - 0s - 10ms/step - accuracy: 0.9997 - loss: 0.0050
8/8 - 0s - 48ms/step
Epoch 1/127


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/127


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/127


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/127


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 3         | nan       | 4.105     | 435.6     | 0.1997    | 0.1543    | 126.6     | 1.093     | 1.608     | 1.171     | 0.00744   | 47.96     | 0.9656    | 5.659     |
Epoch 1/42
57/57 - 3s - 51ms/step - accuracy: 0.6239 - loss: 1.2807
Epoch 2/42
57/57 - 0s - 8ms/step - accuracy: 0.6459 - loss: 1.0506
Epoch 3/42
57/57 - 0s - 7ms/step - accuracy: 0.6566 - loss: 1.0077
Epoch 4/42
57/57 - 0s - 8ms/step - accuracy: 0.6643 - loss: 0.9720
Epoch 5/42
57/57 - 0s - 7ms/step - accuracy: 0.6793 - loss: 0.9331
Epoch 6/42
57/57 - 0s - 8ms/step - accuracy: 0.6858 - loss: 0.9012
Epoch 7/42
57/57 - 0s - 8ms/step - accuracy: 0.7023 - loss: 0.8691
Epoch 8/42
57/57 - 0s - 8ms/step - accuracy: 0.7113 - loss: 0.8457
Epoch 9/42
57/57 - 0s - 7ms/step - accuracy: 0.7206 - loss: 0.8212
Epoch 10/42
57/57 - 0s - 7ms/step - accuracy: 0.7281 - loss: 0.8052
Epoch 11/42
57/57 - 0s - 7ms/step - accuracy: 0.7336 - loss: 0.7881
Epoch 12/42
57/57 - 0s - 7ms/step - accuracy: 0.7397 - loss: 0.7772
Epoch 13/42
57/57 - 0s

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/42
Epoch 1/42


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/42


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/42


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 4         | nan       | 2.742     | 229.3     | 0.6842    | 0.132     | 41.97     | 1.99      | 1.034     | 1.909     | 0.02662   | 36.5      | 0.3117    | 3.64      |
Epoch 1/189
51/51 - 3s - 58ms/step - accuracy: 0.6104 - loss: 1.3919
Epoch 2/189
51/51 - 0s - 7ms/step - accuracy: 0.6648 - loss: 0.9610
Epoch 3/189
51/51 - 0s - 6ms/step - accuracy: 0.6922 - loss: 0.8790
Epoch 4/189
51/51 - 0s - 6ms/step - accuracy: 0.7088 - loss: 0.8206
Epoch 5/189
51/51 - 0s - 6ms/step - accuracy: 0.7158 - loss: 0.7968
Epoch 6/189
51/51 - 0s - 7ms/step - accuracy: 0.7309 - loss: 0.7703
Epoch 7/189
51/51 - 0s - 6ms/step - accuracy: 0.7340 - loss: 0.7519
Epoch 8/189
51/51 - 0s - 7ms/step - accuracy: 0.7355 - loss: 0.7408
Epoch 9/189
51/51 - 0s - 6ms/step - accuracy: 0.7429 - loss: 0.7305
Epoch 10/189
51/51 - 0s - 7ms/step - accuracy: 0.7405 - loss: 0.7237
Epoch 11/189
51/51 - 0s - 6ms/step - accuracy: 0.7525 - loss: 0.7016
Epoch 12/189
51/51 - 0s - 7ms/step - accuracy: 0.7519 - loss: 0.7019
Epoch 13/1

51/51 - 0s - 7ms/step - accuracy: 0.8647 - loss: 0.3869
Epoch 117/189
51/51 - 0s - 6ms/step - accuracy: 0.8683 - loss: 0.3842
Epoch 118/189
51/51 - 0s - 7ms/step - accuracy: 0.8612 - loss: 0.4070
Epoch 119/189
51/51 - 0s - 7ms/step - accuracy: 0.8656 - loss: 0.3883
Epoch 120/189
51/51 - 0s - 7ms/step - accuracy: 0.8701 - loss: 0.3813
Epoch 121/189
51/51 - 0s - 7ms/step - accuracy: 0.8697 - loss: 0.3829
Epoch 122/189
51/51 - 0s - 7ms/step - accuracy: 0.8715 - loss: 0.3797
Epoch 123/189
51/51 - 0s - 7ms/step - accuracy: 0.8687 - loss: 0.3868
Epoch 124/189
51/51 - 0s - 7ms/step - accuracy: 0.8688 - loss: 0.3815
Epoch 125/189
51/51 - 0s - 7ms/step - accuracy: 0.8717 - loss: 0.3797
Epoch 126/189
51/51 - 0s - 7ms/step - accuracy: 0.8747 - loss: 0.3679
Epoch 127/189
51/51 - 0s - 7ms/step - accuracy: 0.8734 - loss: 0.3790
Epoch 128/189
51/51 - 0s - 7ms/step - accuracy: 0.8684 - loss: 0.3803
Epoch 129/189
51/51 - 0s - 7ms/step - accuracy: 0.8724 - loss: 0.3775
Epoch 130/189
51/51 - 0s - 7ms/ste

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/189


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/189


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/189


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 5         | nan       | 4.92      | 255.5     | 0.9696    | 0.2325    | 189.1     | 2.79      | 1.598     | 1.922     | 0.009761  | 17.84     | 0.04523   | 2.277     |
Epoch 1/71
46/46 - 5s - 106ms/step - accuracy: 0.6724 - loss: 0.9650
Epoch 2/71
46/46 - 0s - 10ms/step - accuracy: 0.7547 - loss: 0.7148
Epoch 3/71
46/46 - 0s - 11ms/step - accuracy: 0.7701 - loss: 0.6579
Epoch 4/71
46/46 - 0s - 10ms/step - accuracy: 0.7804 - loss: 0.6203
Epoch 5/71
46/46 - 0s - 11ms/step - accuracy: 0.7820 - loss: 0.6135
Epoch 6/71
46/46 - 0s - 11ms/step - accuracy: 0.7879 - loss: 0.5879
Epoch 7/71
46/46 - 0s - 10ms/step - accuracy: 0.8010 - loss: 0.5579
Epoch 8/71
46/46 - 0s - 10ms/step - accuracy: 0.7991 - loss: 0.5596
Epoch 9/71
46/46 - 0s - 11ms/step - accuracy: 0.8037 - loss: 0.5445
Epoch 10/71
46/46 - 0s - 10ms/step - accuracy: 0.8156 - loss: 0.5233
Epoch 11/71
46/46 - 0s - 10ms/step - accuracy: 0.8170 - loss: 0.5058
Epoch 12/71
46/46 - 0s - 10ms/step - accuracy: 0.8228 - loss: 0.4831
Epoch 13/7

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. Whe

Epoch 1/71
Epoch 1/71


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/71


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 6         | nan       | 3.498     | 281.4     | 0.8287    | 0.107     | 70.57     | 2.085     | 1.141     | 1.802     | 0.008381  | 49.48     | 0.7722    | 1.391     |
Epoch 1/159
29/29 - 2s - 54ms/step - accuracy: 0.5843 - loss: 2.4538
Epoch 2/159
29/29 - 0s - 8ms/step - accuracy: 0.6432 - loss: 1.0411
Epoch 3/159
29/29 - 0s - 8ms/step - accuracy: 0.6591 - loss: 0.9815
Epoch 4/159
29/29 - 0s - 8ms/step - accuracy: 0.6644 - loss: 0.9446
Epoch 5/159
29/29 - 0s - 8ms/step - accuracy: 0.6723 - loss: 0.9255
Epoch 6/159
29/29 - 0s - 8ms/step - accuracy: 0.6743 - loss: 0.9154
Epoch 7/159
29/29 - 0s - 8ms/step - accuracy: 0.6828 - loss: 0.9009
Epoch 8/159
29/29 - 0s - 8ms/step - accuracy: 0.6858 - loss: 0.8936
Epoch 9/159
29/29 - 0s - 8ms/step - accuracy: 0.6840 - loss: 0.8813
Epoch 10/159
29/29 - 0s - 8ms/step - accuracy: 0.6869 - loss: 0.8795
Epoch 11/159
29/29 - 0s - 8ms/step - accuracy: 0.6951 - loss: 0.8541
Epoch 12/159
29/29 - 0s - 8ms/step - accuracy: 0.6980 - loss: 0.8577
Epoch 13/1

29/29 - 0s - 8ms/step - accuracy: 0.7572 - loss: 0.6824
Epoch 117/159
29/29 - 0s - 8ms/step - accuracy: 0.7589 - loss: 0.6766
Epoch 118/159
29/29 - 0s - 9ms/step - accuracy: 0.7516 - loss: 0.6920
Epoch 119/159
29/29 - 0s - 8ms/step - accuracy: 0.7513 - loss: 0.6872
Epoch 120/159
29/29 - 0s - 8ms/step - accuracy: 0.7662 - loss: 0.6683
Epoch 121/159
29/29 - 0s - 8ms/step - accuracy: 0.7608 - loss: 0.6790
Epoch 122/159
29/29 - 0s - 9ms/step - accuracy: 0.7660 - loss: 0.6656
Epoch 123/159
29/29 - 0s - 9ms/step - accuracy: 0.7531 - loss: 0.6858
Epoch 124/159
29/29 - 0s - 8ms/step - accuracy: 0.7632 - loss: 0.6691
Epoch 125/159
29/29 - 0s - 8ms/step - accuracy: 0.7649 - loss: 0.6725
Epoch 126/159
29/29 - 0s - 8ms/step - accuracy: 0.7600 - loss: 0.6738
Epoch 127/159
29/29 - 0s - 8ms/step - accuracy: 0.7633 - loss: 0.6652
Epoch 128/159
29/29 - 0s - 8ms/step - accuracy: 0.7658 - loss: 0.6689
Epoch 129/159
29/29 - 0s - 8ms/step - accuracy: 0.7625 - loss: 0.6686
Epoch 130/159
29/29 - 0s - 8ms/ste

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/159


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/159


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/159


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 7         | nan       | 0.0497    | 444.6     | 0.7069    | 0.2187    | 158.8     | 1.148     | 1.358     | 1.116     | 0.08645   | 34.93     | 0.3309    | 0.4449    |
Epoch 1/180
44/44 - 4s - 101ms/step - accuracy: 0.1962 - loss: 2.5895
Epoch 2/180
44/44 - 0s - 8ms/step - accuracy: 0.6096 - loss: 2.1250
Epoch 3/180
44/44 - 0s - 8ms/step - accuracy: 0.6402 - loss: 1.6814
Epoch 4/180
44/44 - 0s - 8ms/step - accuracy: 0.6419 - loss: 1.3946
Epoch 5/180
44/44 - 0s - 8ms/step - accuracy: 0.6422 - loss: 1.2569
Epoch 6/180
44/44 - 0s - 8ms/step - accuracy: 0.6425 - loss: 1.1863
Epoch 7/180
44/44 - 0s - 8ms/step - accuracy: 0.6431 - loss: 1.1438
Epoch 8/180
44/44 - 0s - 8ms/step - accuracy: 0.6440 - loss: 1.1125
Epoch 9/180
44/44 - 0s - 8ms/step - accuracy: 0.6469 - loss: 1.0872
Epoch 10/180
44/44 - 0s - 8ms/step - accuracy: 0.6509 - loss: 1.0685
Epoch 11/180
44/44 - 0s - 8ms/step - accuracy: 0.6537 - loss: 1.0505
Epoch 12/180
44/44 - 0s - 8ms/step - accuracy: 0.6554 - loss: 1.0356
Epoch 13/

44/44 - 0s - 8ms/step - accuracy: 0.7631 - loss: 0.6900
Epoch 117/180
44/44 - 0s - 8ms/step - accuracy: 0.7628 - loss: 0.6893
Epoch 118/180
44/44 - 0s - 8ms/step - accuracy: 0.7628 - loss: 0.6896
Epoch 119/180
44/44 - 0s - 8ms/step - accuracy: 0.7631 - loss: 0.6881
Epoch 120/180
44/44 - 0s - 8ms/step - accuracy: 0.7652 - loss: 0.6879
Epoch 121/180
44/44 - 0s - 8ms/step - accuracy: 0.7622 - loss: 0.6864
Epoch 122/180
44/44 - 0s - 8ms/step - accuracy: 0.7656 - loss: 0.6854
Epoch 123/180
44/44 - 0s - 8ms/step - accuracy: 0.7627 - loss: 0.6861
Epoch 124/180
44/44 - 0s - 8ms/step - accuracy: 0.7660 - loss: 0.6879
Epoch 125/180
44/44 - 0s - 8ms/step - accuracy: 0.7674 - loss: 0.6821
Epoch 126/180
44/44 - 0s - 8ms/step - accuracy: 0.7610 - loss: 0.6850
Epoch 127/180
44/44 - 0s - 8ms/step - accuracy: 0.7638 - loss: 0.6827
Epoch 128/180
44/44 - 0s - 9ms/step - accuracy: 0.7644 - loss: 0.6814
Epoch 129/180
44/44 - 0s - 8ms/step - accuracy: 0.7633 - loss: 0.6820
Epoch 130/180
44/44 - 0s - 9ms/ste

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/180
Epoch 1/180


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/180


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. Whe

Epoch 1/180
| 8         | nan       | 2.799     | 297.6     | 0.7296    | 0.1913    | 179.7     | 1.944     | 1.12      | 1.713     | 0.07632   | 32.45     | 0.771     | 3.457     |
Epoch 1/26


C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py", line 1501, in fit
    super().fit(X=X, y=y, sample_weight=sample_weight, **kwargs)
  File "C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py", line 770, in fit
    self._fit(
  File "C:\Users\Quinn\anaconda3\Lib\site-packages

40/40 - 4s - 90ms/step - accuracy: 0.5247 - loss: 4.5057
Epoch 2/26
40/40 - 0s - 6ms/step - accuracy: 0.6243 - loss: 1.1173
Epoch 3/26
40/40 - 0s - 6ms/step - accuracy: 0.6217 - loss: 1.0677
Epoch 4/26
40/40 - 0s - 6ms/step - accuracy: 0.6274 - loss: 1.0228
Epoch 5/26
40/40 - 0s - 6ms/step - accuracy: 0.6403 - loss: 1.0003
Epoch 6/26
40/40 - 0s - 6ms/step - accuracy: 0.6411 - loss: 0.9841
Epoch 7/26
40/40 - 0s - 6ms/step - accuracy: 0.6366 - loss: 0.9858
Epoch 8/26
40/40 - 0s - 6ms/step - accuracy: 0.6491 - loss: 0.9577
Epoch 9/26
40/40 - 0s - 6ms/step - accuracy: 0.6576 - loss: 0.9410
Epoch 10/26
40/40 - 0s - 6ms/step - accuracy: 0.6726 - loss: 0.9223
Epoch 11/26
40/40 - 0s - 6ms/step - accuracy: 0.6861 - loss: 0.8779
Epoch 12/26
40/40 - 0s - 6ms/step - accuracy: 0.6995 - loss: 0.8412
Epoch 13/26
40/40 - 0s - 6ms/step - accuracy: 0.7029 - loss: 0.8274
Epoch 14/26
40/40 - 0s - 6ms/step - accuracy: 0.7035 - loss: 0.8111
Epoch 15/26
40/40 - 0s - 6ms/step - accuracy: 0.7169 - loss: 0.7946

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/26


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/26


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/26


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 9         | nan       | 4.705     | 328.3     | 0.02542   | 0.03237   | 25.66     | 2.273     | 1.314     | 1.509     | 0.09085   | 19.97     | 0.4104    | 5.289     |
Epoch 1/187
58/58 - 3s - 55ms/step - accuracy: 0.6278 - loss: 1.1944
Epoch 2/187
58/58 - 0s - 7ms/step - accuracy: 0.6677 - loss: 1.0028
Epoch 3/187
58/58 - 0s - 7ms/step - accuracy: 0.7117 - loss: 0.9080
Epoch 4/187
58/58 - 0s - 7ms/step - accuracy: 0.7162 - loss: 0.8663
Epoch 5/187
58/58 - 0s - 7ms/step - accuracy: 0.7259 - loss: 0.8258
Epoch 6/187
58/58 - 0s - 7ms/step - accuracy: 0.7347 - loss: 0.7823
Epoch 7/187
58/58 - 0s - 7ms/step - accuracy: 0.7432 - loss: 0.7474
Epoch 8/187
58/58 - 0s - 7ms/step - accuracy: 0.7516 - loss: 0.7268
Epoch 9/187
58/58 - 0s - 6ms/step - accuracy: 0.7593 - loss: 0.7083
Epoch 10/187
58/58 - 0s - 7ms/step - accuracy: 0.7624 - loss: 0.6885
Epoch 11/187
58/58 - 0s - 7ms/step - accuracy: 0.7652 - loss: 0.6808
Epoch 12/187
58/58 - 0s - 7ms/step - accuracy: 0.7707 - loss: 0.6674
Epoch 13/1

58/58 - 0s - 7ms/step - accuracy: 0.8453 - loss: 0.4169
Epoch 117/187
58/58 - 0s - 7ms/step - accuracy: 0.8447 - loss: 0.4209
Epoch 118/187
58/58 - 0s - 7ms/step - accuracy: 0.8415 - loss: 0.4254
Epoch 119/187
58/58 - 0s - 7ms/step - accuracy: 0.8421 - loss: 0.4220
Epoch 120/187
58/58 - 0s - 7ms/step - accuracy: 0.8431 - loss: 0.4199
Epoch 121/187
58/58 - 0s - 7ms/step - accuracy: 0.8376 - loss: 0.4265
Epoch 122/187
58/58 - 0s - 7ms/step - accuracy: 0.8439 - loss: 0.4176
Epoch 123/187
58/58 - 0s - 7ms/step - accuracy: 0.8431 - loss: 0.4146
Epoch 124/187
58/58 - 0s - 7ms/step - accuracy: 0.8452 - loss: 0.4170
Epoch 125/187
58/58 - 0s - 7ms/step - accuracy: 0.8457 - loss: 0.4157
Epoch 126/187
58/58 - 0s - 7ms/step - accuracy: 0.8440 - loss: 0.4205
Epoch 127/187
58/58 - 0s - 7ms/step - accuracy: 0.8420 - loss: 0.4133
Epoch 128/187
58/58 - 0s - 7ms/step - accuracy: 0.8477 - loss: 0.4060
Epoch 129/187
58/58 - 0s - 7ms/step - accuracy: 0.8459 - loss: 0.4149
Epoch 130/187
58/58 - 0s - 7ms/ste

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/187


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/187


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/187


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 10        | nan       | 2.059     | 223.1     | 0.2898    | 0.04837   | 187.3     | 2.616     | 1.633     | 1.871     | 0.08056   | 17.46     | 0.8926    | 3.775     |
Epoch 1/61
28/28 - 3s - 103ms/step - accuracy: 0.6423 - loss: nan
Epoch 2/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 3/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 4/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 5/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 6/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 7/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 8/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 9/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 10/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 11/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 12/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - loss: nan
Epoch 13/61
28/28 - 0s - 9ms/step - accuracy: 0.6423 - lo

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/61


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/61


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/61


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 11        | nan       | 7.267     | 468.8     | 0.318     | 0.03302   | 61.03     | 1.854     | 1.818     | 1.861     | 0.001688  | 30.43     | 0.4174    | 1.555     |
Epoch 1/113
43/43 - 3s - 61ms/step - accuracy: 0.6044 - loss: 1.5147
Epoch 2/113
43/43 - 0s - 7ms/step - accuracy: 0.6240 - loss: 1.1614
Epoch 3/113
43/43 - 0s - 7ms/step - accuracy: 0.6352 - loss: 1.0634
Epoch 4/113
43/43 - 0s - 7ms/step - accuracy: 0.6416 - loss: 1.0166
Epoch 5/113
43/43 - 0s - 7ms/step - accuracy: 0.6478 - loss: 0.9960
Epoch 6/113
43/43 - 0s - 7ms/step - accuracy: 0.6523 - loss: 0.9774
Epoch 7/113
43/43 - 0s - 7ms/step - accuracy: 0.6649 - loss: 0.9470
Epoch 8/113
43/43 - 0s - 7ms/step - accuracy: 0.6647 - loss: 0.9345
Epoch 9/113
43/43 - 0s - 8ms/step - accuracy: 0.6657 - loss: 0.9209
Epoch 10/113
43/43 - 0s - 7ms/step - accuracy: 0.6733 - loss: 0.9007
Epoch 11/113
43/43 - 0s - 7ms/step - accuracy: 0.6736 - loss: 0.8950
Epoch 12/113
43/43 - 0s - 7ms/step - accuracy: 0.6780 - loss: 0.8726
Epoch 13/1

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/113


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/113


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/113


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 12        | nan       | 1.079     | 301.3     | 0.9429    | 0.09696   | 113.4     | 2.406     | 1.364     | 1.972     | 0.09628   | 20.07     | 0.4972    | 2.106     |
Epoch 1/29
61/61 - 3s - 57ms/step - accuracy: 0.6524 - loss: 1.0338
Epoch 2/29
61/61 - 0s - 8ms/step - accuracy: 0.7075 - loss: 0.8324
Epoch 3/29
61/61 - 0s - 7ms/step - accuracy: 0.7165 - loss: 0.7955
Epoch 4/29
61/61 - 0s - 7ms/step - accuracy: 0.7249 - loss: 0.7705
Epoch 5/29
61/61 - 0s - 7ms/step - accuracy: 0.7402 - loss: 0.7336
Epoch 6/29
61/61 - 0s - 7ms/step - accuracy: 0.7436 - loss: 0.7272
Epoch 7/29
61/61 - 0s - 8ms/step - accuracy: 0.7528 - loss: 0.6999
Epoch 8/29
61/61 - 0s - 8ms/step - accuracy: 0.7453 - loss: 0.7030
Epoch 9/29
61/61 - 0s - 7ms/step - accuracy: 0.7520 - loss: 0.6983
Epoch 10/29
61/61 - 0s - 8ms/step - accuracy: 0.7595 - loss: 0.6768
Epoch 11/29
61/61 - 0s - 7ms/step - accuracy: 0.7610 - loss: 0.6585
Epoch 12/29
61/61 - 0s - 8ms/step - accuracy: 0.7591 - loss: 0.6719
Epoch 13/29
61/61 - 0s

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/29


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. Whe

Epoch 1/29
Epoch 1/29


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 13        | nan       | 2.564     | 211.1     | 0.6096    | 0.1508    | 29.27     | 1.557     | 1.908     | 1.24      | 0.01534   | 29.58     | 0.9857    | 1.694     |
Epoch 1/86
31/31 - 5s - 153ms/step - accuracy: 0.6244 - loss: 1.2123
Epoch 2/86
31/31 - 0s - 11ms/step - accuracy: 0.7337 - loss: 0.7840
Epoch 3/86
31/31 - 0s - 11ms/step - accuracy: 0.7520 - loss: 0.7268
Epoch 4/86
31/31 - 0s - 11ms/step - accuracy: 0.7594 - loss: 0.6926
Epoch 5/86
31/31 - 0s - 11ms/step - accuracy: 0.7575 - loss: 0.6824
Epoch 6/86
31/31 - 0s - 10ms/step - accuracy: 0.7778 - loss: 0.6322
Epoch 7/86
31/31 - 0s - 11ms/step - accuracy: 0.7831 - loss: 0.6049
Epoch 8/86
31/31 - 0s - 11ms/step - accuracy: 0.7781 - loss: 0.6288
Epoch 9/86
31/31 - 0s - 11ms/step - accuracy: 0.7799 - loss: 0.6122
Epoch 10/86
31/31 - 0s - 11ms/step - accuracy: 0.7850 - loss: 0.5993
Epoch 11/86
31/31 - 0s - 11ms/step - accuracy: 0.7887 - loss: 0.5905
Epoch 12/86
31/31 - 0s - 11ms/step - accuracy: 0.7976 - loss: 0.5597
Epoch 13/8

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/86
Epoch 1/86


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/86


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/86


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------

| 14        | nan       | 6.049     | 428.5     | 0.2376    | 0.2185    | 86.2      | 2.265     | 1.634     | 1.536     | 0.009939  | 43.41     | 0.3208    | 1.306     |
Epoch 1/112
35/35 - 5s - 134ms/step - accuracy: 0.6209 - loss: 1.2029
Epoch 2/112
35/35 - 0s - 9ms/step - accuracy: 0.7106 - loss: 0.8357
Epoch 3/112
35/35 - 0s - 9ms/step - accuracy: 0.7165 - loss: 0.8027
Epoch 4/112
35/35 - 0s - 9ms/step - accuracy: 0.7330 - loss: 0.7388
Epoch 5/112
35/35 - 0s - 9ms/step - accuracy: 0.7388 - loss: 0.7154
Epoch 6/112
35/35 - 0s - 9ms/step - accuracy: 0.7271 - loss: 0.7486
Epoch 7/112
35/35 - 0s - 10ms/step - accuracy: 0.7477 - loss: 0.7028
Epoch 8/112
35/35 - 0s - 9ms/step - accuracy: 0.7559 - loss: 0.6881
Epoch 9/112
35/35 - 0s - 9ms/step - accuracy: 0.7681 - loss: 0.6675
Epoch 10/112
35/35 - 0s - 9ms/step - accuracy: 0.7764 - loss: 0.6327
Epoch 11/112
35/35 - 0s - 8ms/step - accuracy: 0.6760 - loss: 0.9199
Epoch 12/112
35/35 - 0s - 9ms/step - accuracy: 0.7152 - loss: 0.7725
Epoch 13

C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/112


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. Whe

Epoch 1/112


C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/112


C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
4 fits failed out of a total of 5.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Quinn\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py", line 1501, in fit
    super().fit(X=X, y=y, sample_weight=sample_weight, **kwargs)
  File "C:\Users\Quinn\anaconda3\Lib\site-packages\scikeras\wrappers.py", line 770, in fit
    self._fit(
  File "C:\Users\Quinn\anaconda3\Lib\site-packages

| 15        | nan       | 0.367     | 377.3     | 0.6776    | 0.004976  | 112.2     | 1.453     | 1.645     | 1.174     | 0.0694    | 25.47     | 0.9367    | 0.9626    |


ValueError: Input y contains NaN.

In [18]:
# Display the best parameters
optimum = nn_opt.max['params']
learning_rate = optimum['learning_rate']
activationL = ['relu', 'sigmoid', 'softplus', 'softsign', 'tanh', 'selu',
               'elu', 'exponential', LeakyReLU,'relu']
optimum['activation'] = activationL[round(optimum['activation'])]
optimum['batch_size'] = round(optimum['batch_size']) 
optimum['epochs'] = round(optimum['epochs'])
optimum['layers1'] = round(optimum['layers1'])
optimum['layers2'] = round(optimum['layers2'])
optimum['neurons'] = round(optimum['neurons'])
optimizerL = ['Adam', 'SGD', 'RMSprop', 'Adadelta', 'Adagrad', 'Adamax', 'Nadam', 'Ftrl','Adam']
optimizerD= {'Adam':Adam(learning_rate=learning_rate), 'SGD':SGD(learning_rate=learning_rate),
             'RMSprop':RMSprop(learning_rate=learning_rate), 'Adadelta':Adadelta(learning_rate=learning_rate),
             'Adagrad':Adagrad(learning_rate=learning_rate), 'Adamax':Adamax(learning_rate=learning_rate),
             'Nadam':Nadam(learning_rate=learning_rate), 'Ftrl':Ftrl(learning_rate=learning_rate)}
optimum['optimizer'] = optimizerD[optimizerL[round(optimum['optimizer'])]]
optimum

{'activation': 'softsign',
 'batch_size': 485,
 'dropout': 0.7319939418114051,
 'dropout_rate': 0.17959754525911098,
 'epochs': 48,
 'kernel': 1.3119890406724053,
 'layers1': 1,
 'layers2': 2,
 'learning_rate': 0.06051038616257767,
 'neurons': 38,
 'normalization': 0.020584494295802447,
 'optimizer': <keras.src.optimizers.ftrl.Ftrl at 0x17c5ba25d10>}

## Create a Keras layered model

In [19]:
# Turn y_test into one-hot format.

y_test=tf.keras.utils.to_categorical(y_test)
print(y_test)

[[1. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


## 04. CNN Model 

In [20]:
# Define CNN parameters with optimums returned by Bayesian Optimization.

timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
activation = 'softsign'
batch_size = optimum['batch_size']
dropout = optimum['dropout']
dropout_rate = optimum['dropout_rate']
epochs = optimum['epochs']
kernel = 1 # Running optimum['kernel'] threw a ValueError on the CNN model below that stated the kernel_size argument must be a whole integer. I rounded the 'kernel' output above down to 1.
layers1 = 1 
layers2 = 2 
learning_rate = optimum['learning_rate']
neurons = 38 # Must be a whole integer for the CNN model
normalization = optimum['normalization']
optimizer = optimum['optimizer']

In [21]:
# Create CNN model with above optimum parameters.

model = Sequential()
model.add(Conv1D(neurons, kernel_size=kernel,activation=activation, input_shape=(timesteps, input_dim)))
if normalization > 0.5:
    model.add(BatchNormalization())
for i in range(layers1):
    model.add(Dense(neurons, activation=activation)) #(neurons, activation=activation))
if dropout > 0.5:
    model.add(Dropout(dropout_rate, seed=123))
for i in range(layers2):
    model.add(Dense(neurons, activation=activation))
model.add(MaxPooling1D())
model.add(Flatten())
model.add(Dense(n_classes, activation='softmax')) #sigmoid softmax
model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

C:\Users\Quinn\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [22]:
# Run model.

model.fit(X_train,y_train,batch_size=485, epochs=48,verbose=0)
acc = model.evaluate(X_train,y_train)
print('Loss:', acc[0], 'Accuracy', acc[1])

503/503 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7131 - loss: 0.8439
Loss: 0.8411258459091187 Accuracy 0.7162153720855713


#### Notes: smaller loss and accuracy increased

### Confusion matrix

In [23]:
# Define cities.
cities = [
    "Basel",
    "Belgrade",
    "Budapest",
    "Debilt",
    "Dusseldorf",
    "Heathrow",
    "Kassel",
    "Ljubljana",
    "Maastricht",
    "Madrid",
    "Munchen",
    "Oslo",
    "Sonnblick",
    "Stockholm",
    "Valentia"
]

In [24]:
# Define confusion matrix to see results.

def confusion_matrix(Y_true, Y_pred):
    Y_true = pd.Series([cities[y] for y in np.argmax(Y_true, axis=1)])
    Y_pred = pd.Series([cities[y] for y in np.argmax(Y_pred, axis=1)])

    return pd.crosstab(Y_true, Y_pred, rownames=['True'], colnames=['Pred'])

In [25]:
# Print results.

print(confusion_matrix(y_test, model.predict(X_test)))

216/216 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step
Pred        Basel  Belgrade  Madrid
True                               
Basel        4220       222       5
Belgrade      664       640       1
Budapest      164        74       7
Debilt         79        16       1
Dusseldorf     23         8       1
Heathrow       71        14      14
Kassel         10         3       0
Ljubljana      55         8       8
Maastricht      9         0       0
Madrid        474        51      21
Munchen         8         0       0
Oslo            4         2       1
Stockholm       1         3       0
Valentia        3         0       0


####  The model only recognized 3 cities. The accuracy score for the given confusion matrix is approximately 71.62%